## fix unit_price in order_lines (from total_paid of orders)

In [28]:
import pandas as pd

ord_l = pd.read_csv('orderlines.csv')
ords = pd.read_csv('orders.csv')

In [29]:
ord_l_df = ord_l.copy()
ords_df = ords.copy()

In [30]:
ord_l_df.info()
ords_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
me

### remove missing values

In [33]:
ords_df.total_paid.isna().sum()
ords_df.isna().sum()

order_id        0
created_date    0
total_paid      5
state           0
dtype: int64

In [34]:
ords_df = ords_df.loc[~ords_df.total_paid.isna(), :]
ords_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226904 non-null  int64  
 1   created_date  226904 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226904 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 8.7+ MB


In [35]:
ords_df["created_date"] = pd.to_datetime(ords_df["created_date"])
ords_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 8.7+ MB


## Orderlines

In [36]:
ord_l_df.info()
ord_l_df["date"] = pd.to_datetime(ord_l_df["date"])
ord_l_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 n

## Orderlines unit_price
### they show some courrpted prices:

In [37]:
ord_l_df.unit_price.head(10)

0       18.99
1      399.00
2      474.05
3       68.39
4       23.74
5      231.79
6    1.137.99
7       47.49
8       18.99
9       72.19
Name: unit_price, dtype: object

In [38]:
cond_two_dots = ord_l_df.unit_price.str.contains("\d+\.\d+\.\d+")
cond_two_dots.head(10)

0    False
1    False
2    False
3    False
4    False
5    False
6     True
7    False
8    False
9    False
Name: unit_price, dtype: bool

In [39]:
ord_l_df[cond_two_dots]

id  id_order  product_id  product_quantity      sku unit_price  \
6       1119115    299544           0                 1  APP1582   1.137.99   
11      1119126    299549           0                 1  PAC0929   2.565.99   
15      1119131    299553           0                 1  APP1854   3.278.99   
43      1119195    299582           0                 1  PAC0961   2.616.99   
59      1119214    299596           0                 1  PAC1599   2.873.99   
...         ...       ...         ...               ...      ...        ...   
293862  1649999    452946           0                 1  APP2075   2.999.00   
293887  1650045    527321           0                 1  PAC2148   3.497.00   
293889  1650050    527324           0                 1  PAC2117   3.075.00   
293911  1650088    527342           0                 1  APP2492   1.329.00   
293936  1650135    527364           0                 1  APP2491   1.329.00   

                      date  
6      2017-01-01 01:17:21  
11     2017-01-01 02:07:42  
15     2017-01-01 02:14:47  
43     2017-01-01 08:54:00  
59     2017-01-01 09:53:11  
...                    ...  
293862 2018-03-14 13:03:33  
293887 2018-03-14 13:10:15  
293889 2018-03-14 13:10:56  
293911 2018-03-14 13:24:51  
293936 2018-03-14 13:36:13  

[36169 rows x 7 columns]

## orderlines without any id_order duplicates, excluding also first occurrence, and copy

In [44]:
ord_l_df.info()
print('\n', ord_l_df.drop_duplicates(subset=['id_order'], keep=False)['id_order'].count(), 'rows of duplicates')
print(ord_l_df.shape[0], 'total rows')
print(ord_l_df.shape[0] - ord_l_df.drop_duplicates(subset=['id_order'], keep=False)['id_order'].count(), 'discarded rows')
#ord_l_df.loc[~ord_l_df.id_order.duplicated()]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  object        
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 15.7+ MB

 156141 rows of duplicates
293983 total rows
137842 discarded rows


In [46]:
# new df without duplicates:
ord_l_df_no_dupl = ord_l_df.drop_duplicates(subset=['id_order'], keep=False)
ord_l_df_no_dupl.nunique()
ord_l_df_no_dupl.info()

<class 'pandas.core.frame.DataFrame'>
Index: 156141 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                156141 non-null  int64         
 1   id_order          156141 non-null  int64         
 2   product_id        156141 non-null  int64         
 3   product_quantity  156141 non-null  int64         
 4   sku               156141 non-null  object        
 5   unit_price        156141 non-null  object        
 6   date              156141 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 9.5+ MB


In [47]:
# count number of corrupted prices in new df without duplicates:
ord_l_df_no_dupl.loc[ord_l_df_no_dupl.unit_price.str.contains("\d+\.\d+\.\d+"), 'unit_price'].count()

17555

In [48]:
del ord_l_df_no_dupl['product_id']

In [58]:
# sort ordersL with corrupted prices by the product_quantity:
ord_l_df_no_dupl.loc[ord_l_df_no_dupl.unit_price.str.contains("\d+\.\d+\.\d+")].sort_values(by='product_quantity')

id  id_order  product_quantity      sku unit_price  \
6       1119115    299544                 1  APP1582   1.137.99   
198419  1491031    460757                 1  APP2089   1.979.00   
198474  1491133    460806                 1  APP2373   2.209.01   
198503  1491181    460832                 1  APP2488   1.067.00   
198526  1491214    460851                 1  APP2488   1.067.00   
...         ...       ...               ...      ...        ...   
151704  1419940    430951                21  APP2355   4.717.59   
96238   1303180    382381                22  APP2264   2.650.00   
46159   1214475    339637                25  APP1195   1.189.99   
13851   1152175    311950                35  WAC0030   1.699.99   
39923   1203123    334310                98  APP1632   1.019.00   

                      date  
6      2017-01-01 01:17:21  
198419 2017-12-12 14:21:51  
198474 2017-12-12 15:32:02  
198503 2017-12-12 16:11:43  
198526 2017-12-12 16:40:01  
...                    ...  
151704 2017-11-22 02:00:55  
96238  2017-07-26 16:09:57  
46159  2017-03-27 12:43:15  
13851  2017-01-19 09:45:05  
39923  2017-03-12 12:59:10  

[17555 rows x 6 columns]

In [59]:
# filter ordersL with corrupted prices & product_quantity==1:
ord_l_df_no_dupl.loc[(
            (ord_l_df_no_dupl.unit_price.str.contains("\d+\.\d+\.\d+")) &
            (ord_l_df_no_dupl['product_quantity']==1) ),
            'id_order'
                    ]

6         299544
11        299549
15        299553
43        299582
105       299629
           ...  
293838    527286
293887    527321
293889    527324
293911    527342
293936    527364
Name: id_order, Length: 17096, dtype: int64

In [60]:
# list of id orders: ONLY from ordersL with corrupted prices that also have product_quantity==1:
corrupted_ids = ord_l_df_no_dupl.loc[(
                            (ord_l_df_no_dupl.unit_price.str.contains("\d+\.\d+\.\d+")) &
                            (ord_l_df_no_dupl['product_quantity']==1) ),
                            'id_order'
                                    ]

## copy of orders

In [61]:
# copy of orders filtered by the ids with ONLY corrupted prices from orderL
ords_df_cp = ords_df[ords_df['order_id'].isin(corrupted_ids)].copy()
ords_df_cp

order_id        created_date  total_paid            state
31        257847 2017-11-23 23:46:54     1367.11        Completed
87        279755 2017-01-08 22:25:24     2616.99        Completed
115       287876 2017-03-07 16:54:18     2252.24          Pending
157       296430 2017-03-27 23:20:39     2673.99        Cancelled
163       296986 2017-03-30 12:24:58     4736.99        Cancelled
...          ...                 ...         ...              ...
226793    527286 2018-03-14 12:55:14     1165.99      Place Order
226828    527321 2018-03-14 13:16:30     3503.99          Pending
226831    527324 2018-03-14 13:10:56     3075.00  Shopping Basket
226849    527342 2018-03-14 13:24:51     1335.99      Place Order
226871    527364 2018-03-14 13:36:13     1329.00  Shopping Basket

[17089 rows x 4 columns]

## Merge of the right total_paid into the original orderlines df (ord_l_df):

In [24]:
ord_l_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  object        
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 15.7+ MB


In [25]:
ord_l_df_merge = ord_l_df.merge(
                                ords_df_cp,
                                how='left',
                                left_on='id_order',
                                right_on='order_id'
                                )
ord_l_df_merge.info()
ord_l_df_merge.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  object        
 6   date              293983 non-null  datetime64[ns]
 7   order_id          17089 non-null   float64       
 8   created_date      17089 non-null   datetime64[ns]
 9   total_paid        17089 non-null   float64       
 10  state             17089 non-null   object        
dtypes: datetime64[ns](2), float64(2), int64(4), object(3)
memory usage: 24.7+ MB


id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   
5  1119114    295310           0                10  WDT0249     231.79   
6  1119115    299544           0                 1  APP1582   1.137.99   
7  1119116    299545           0                 1  OWC0100      47.49   
8  1119119    299546           0                 1  IOT0014      18.99   
9  1119120    295347           0                 1  APP0700      72.19   

                 date  order_id        created_date  total_paid  \
0 2017-01-01 00:07:19       NaN                 NaT         NaN   
1 2017-01-01 00:19:45       NaN                 NaT         NaN   
2 2017-01-01 00:20:57       NaN                 NaT         NaN   
3 2017-01-01 00:51:40       NaN                 NaT         NaN   
4 2017-01-01 01:06:38       NaN                 NaT         NaN   
5 2017-01-01 01:14:27       NaN                 NaT         NaN   
6 2017-01-01 01:17:21  299544.0 2017-01-01 01:17:21     1137.99   
7 2017-01-01 01:46:16       NaN                 NaT         NaN   
8 2017-01-01 01:50:34       NaN                 NaT         NaN   
9 2017-01-01 01:54:11       NaN                 NaT         NaN   

             state  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
5              NaN  
6  Shopping Basket  
7              NaN  
8              NaN  
9              NaN

In [26]:
ord_l_df_merge[ord_l_df_merge['id_order'].isin(corrupted_ids)]

id  id_order  product_id  product_quantity      sku unit_price  \
6       1119115    299544           0                 1  APP1582   1.137.99   
11      1119126    299549           0                 1  PAC0929   2.565.99   
15      1119131    299553           0                 1  APP1854   3.278.99   
43      1119195    299582           0                 1  PAC0961   2.616.99   
105     1119294    299629           0                 1  QNA0175   2.054.99   
...         ...       ...         ...               ...      ...        ...   
293838  1649954    527286           0                 1  APP2490   1.159.00   
293887  1650045    527321           0                 1  PAC2148   3.497.00   
293889  1650050    527324           0                 1  PAC2117   3.075.00   
293911  1650088    527342           0                 1  APP2492   1.329.00   
293936  1650135    527364           0                 1  APP2491   1.329.00   

                      date  order_id        created_date  total_paid  \
6      2017-01-01 01:17:21  299544.0 2017-01-01 01:17:21     1137.99   
11     2017-01-01 02:07:42  299549.0 2017-01-02 10:00:20     2565.99   
15     2017-01-01 02:14:47  299553.0 2017-01-01 02:14:47     3278.99   
43     2017-01-01 08:54:00  299582.0 2017-01-01 09:16:18     2616.99   
105    2017-01-01 11:39:13  299629.0 2017-01-01 11:39:13     2054.99   
...                    ...       ...                 ...         ...   
293838 2018-03-14 12:55:14  527286.0 2018-03-14 12:55:14     1165.99   
293887 2018-03-14 13:10:15  527321.0 2018-03-14 13:16:30     3503.99   
293889 2018-03-14 13:10:56  527324.0 2018-03-14 13:10:56     3075.00   
293911 2018-03-14 13:24:51  527342.0 2018-03-14 13:24:51     1335.99   
293936 2018-03-14 13:36:13  527364.0 2018-03-14 13:36:13     1329.00   

                  state  
6       Shopping Basket  
11            Completed  
15          Place Order  
43            Cancelled  
105     Shopping Basket  
...                 ...  
293838      Place Order  
293887          Pending  
293889  Shopping Basket  
293911      Place Order  
293936  Shopping Basket  

[17096 rows x 11 columns]

In [27]:
ord_l_df_merge[(ord_l_df_merge['id_order'].isin(corrupted_ids)) & (ord_l_df_merge['state']=='Completed')]

id  id_order  product_id  product_quantity      sku unit_price  \
11      1119126    299549           0                 1  PAC0929   2.565.99   
159     1119405    299664           0                 1  APP0958   1.351.99   
230     1119549    299744           0                 1  APP1810   1.584.99   
386     1119836    299901           0                 1  APP1198   1.986.99   
755     1120543    300253           0                 1  APP1267   1.949.99   
...         ...       ...         ...               ...      ...        ...   
292716  1647977    526298           0                 1  APP2490   1.159.00   
292804  1648114    526357           0                 1  APP2078   1.399.00   
292839  1648173    526380           0                 1  APP2087   1.430.00   
293001  1648439    526505           0                 1  APP2064   1.415.00   
293174  1648728    526648           0                 1  APP2073   1.199.00   

                      date  order_id        created_date  total_paid  \
11     2017-01-01 02:07:42  299549.0 2017-01-02 10:00:20     2565.99   
159    2017-01-01 12:57:21  299664.0 2017-01-01 20:08:09     1351.99   
230    2017-01-01 15:19:42  299744.0 2017-01-02 09:22:07     1584.99   
386    2017-01-01 19:32:29  299901.0 2017-01-01 19:47:54     1986.99   
755    2017-01-02 07:56:57  300253.0 2017-01-02 07:59:23     1949.99   
...                    ...       ...                 ...         ...   
292716 2018-03-13 09:57:44  526298.0 2018-03-13 11:17:33     1165.99   
292804 2018-03-13 11:33:52  526357.0 2018-03-13 12:01:26     1405.99   
292839 2018-03-13 12:14:52  526380.0 2018-03-13 12:16:06     1436.99   
293001 2018-03-13 15:50:13  526505.0 2018-03-13 15:56:14     1421.99   
293174 2018-03-13 20:40:12  526648.0 2018-03-13 21:07:19     1205.99   

            state  
11      Completed  
159     Completed  
230     Completed  
386     Completed  
755     Completed  
...           ...  
292716  Completed  
292804  Completed  
292839  Completed  
293001  Completed  
293174  Completed  

[2367 rows x 11 columns]

In [122]:
ord_l_df_merge[~ord_l_df_merge['total_paid'].isna()]

id  id_order  product_id  product_quantity      sku unit_price  \
6       1119115    299544           0                 1  APP1582   1.137.99   
11      1119126    299549           0                 1  PAC0929   2.565.99   
15      1119131    299553           0                 1  APP1854   3.278.99   
43      1119195    299582           0                 1  PAC0961   2.616.99   
105     1119294    299629           0                 1  QNA0175   2.054.99   
...         ...       ...         ...               ...      ...        ...   
293838  1649954    527286           0                 1  APP2490   1.159.00   
293887  1650045    527321           0                 1  PAC2148   3.497.00   
293889  1650050    527324           0                 1  PAC2117   3.075.00   
293911  1650088    527342           0                 1  APP2492   1.329.00   
293936  1650135    527364           0                 1  APP2491   1.329.00   

                      date  order_id        created_date  total_paid  \
6      2017-01-01 01:17:21  299544.0 2017-01-01 01:17:21     1137.99   
11     2017-01-01 02:07:42  299549.0 2017-01-02 10:00:20     2565.99   
15     2017-01-01 02:14:47  299553.0 2017-01-01 02:14:47     3278.99   
43     2017-01-01 08:54:00  299582.0 2017-01-01 09:16:18     2616.99   
105    2017-01-01 11:39:13  299629.0 2017-01-01 11:39:13     2054.99   
...                    ...       ...                 ...         ...   
293838 2018-03-14 12:55:14  527286.0 2018-03-14 12:55:14     1165.99   
293887 2018-03-14 13:10:15  527321.0 2018-03-14 13:16:30     3503.99   
293889 2018-03-14 13:10:56  527324.0 2018-03-14 13:10:56     3075.00   
293911 2018-03-14 13:24:51  527342.0 2018-03-14 13:24:51     1335.99   
293936 2018-03-14 13:36:13  527364.0 2018-03-14 13:36:13     1329.00   

                  state  
6       Shopping Basket  
11            Completed  
15          Place Order  
43            Cancelled  
105     Shopping Basket  
...                 ...  
293838      Place Order  
293887          Pending  
293889  Shopping Basket  
293911      Place Order  
293936  Shopping Basket  

[17089 rows x 11 columns]

In [124]:
ord_l_df_merge[ord_l_df_merge['id_order'].isin(corrupted_ids) & ord_l_df_merge['total_paid'].isna()]

id  id_order  product_id  product_quantity      sku unit_price  \
382     1119824    295154           0                 1  APP1581   1.024.99   
1270    1121682    298608           0                 1  APP1880   2.825.61   
38730   1200904    291213           0                 1  APP1821   1.884.99   
77896   1270117    259150           0                 1  APP2191   1.766.00   
115888  1350875    283110           0                 1  APP2077   3.113.60   
176040  1461236    449596           0                 1  APP2191   1.722.59   
197877  1490090    299397           0                 1  APP2078   1.399.00   

                      date  order_id created_date  total_paid state  
382    2017-01-01 19:27:54       NaN          NaT         NaN   NaN  
1270   2017-01-02 16:39:29       NaN          NaT         NaN   NaN  
38730  2017-03-08 16:22:42       NaN          NaT         NaN   NaN  
77896  2017-06-18 10:01:49       NaN          NaT         NaN   NaN  
115888 2017-09-12 16:22:35       NaN          NaT         NaN   NaN  
176040 2017-11-27 21:49:21       NaN          NaT         NaN   NaN  
197877 2017-12-11 21:48:53       NaN          NaT         NaN   NaN

In [133]:
ord_l_df_merge.head(7)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   
5  1119114    295310           0                10  WDT0249     231.79   
6  1119115    299544           0                 1  APP1582   1.137.99   

                 date  order_id        created_date  total_paid  \
0 2017-01-01 00:07:19       NaN                 NaT         NaN   
1 2017-01-01 00:19:45       NaN                 NaT         NaN   
2 2017-01-01 00:20:57       NaN                 NaT         NaN   
3 2017-01-01 00:51:40       NaN                 NaT         NaN   
4 2017-01-01 01:06:38       NaN                 NaT         NaN   
5 2017-01-01 01:14:27       NaN                 NaT         NaN   
6 2017-01-01 01:17:21  299544.0 2017-01-01 01:17:21     1137.99   

             state  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
5              NaN  
6  Shopping Basket

In [135]:
ord_l_df_merge['product_id'].value_counts()

product_id
0    293983
Name: count, dtype: int64

In [137]:
del ord_l_df_merge['product_id']

In [138]:
ord_l_df_merge['new_unit_price'] = ord_l_df_merge['total_paid'].fillna(ord_l_df_merge['unit_price'])

In [141]:
ord_l_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  object        
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 15.7+ MB


In [142]:
ord_l_df_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_quantity  293983 non-null  int64         
 3   sku               293983 non-null  object        
 4   unit_price        293983 non-null  object        
 5   date              293983 non-null  datetime64[ns]
 6   order_id          17089 non-null   float64       
 7   created_date      17089 non-null   datetime64[ns]
 8   total_paid        17089 non-null   float64       
 9   state             17089 non-null   object        
 10  new_unit_price    293983 non-null  object        
dtypes: datetime64[ns](2), float64(2), int64(3), object(4)
memory usage: 24.7+ MB


In [144]:
del ord_l_df_merge['order_id']
del ord_l_df_merge['created_date']
del ord_l_df_merge['total_paid']
del ord_l_df_merge['state']

In [145]:
ord_l_df_merge.head(10)

id  id_order  product_quantity      sku unit_price  \
0  1119109    299539                 1  OTT0133      18.99   
1  1119110    299540                 1  LGE0043     399.00   
2  1119111    299541                 1  PAR0071     474.05   
3  1119112    299542                 1  WDT0315      68.39   
4  1119113    299543                 1  JBL0104      23.74   
5  1119114    295310                10  WDT0249     231.79   
6  1119115    299544                 1  APP1582   1.137.99   
7  1119116    299545                 1  OWC0100      47.49   
8  1119119    299546                 1  IOT0014      18.99   
9  1119120    295347                 1  APP0700      72.19   

                 date new_unit_price  
0 2017-01-01 00:07:19          18.99  
1 2017-01-01 00:19:45         399.00  
2 2017-01-01 00:20:57         474.05  
3 2017-01-01 00:51:40          68.39  
4 2017-01-01 01:06:38          23.74  
5 2017-01-01 01:14:27         231.79  
6 2017-01-01 01:17:21        1137.99  
7 2017-01-01 01:46:16          47.49  
8 2017-01-01 01:50:34          18.99  
9 2017-01-01 01:54:11          72.19

In [195]:
ord_l_df_merge['new_unit_price'].isna().sum()

0

In [196]:
ord_l_df_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_quantity  293983 non-null  int64         
 3   sku               293983 non-null  object        
 4   unit_price        293983 non-null  object        
 5   date              293983 non-null  datetime64[ns]
 6   new_unit_price    293983 non-null  object        
dtypes: datetime64[ns](1), int64(3), object(3)
memory usage: 15.7+ MB


In [165]:
ord_l_df_merge.loc[ord_l_df_merge['new_unit_price'].astype(str).str.contains("\d+\.\d+\.\d+")]

id  id_order  product_quantity      sku unit_price  \
59      1119214    299596                 1  PAC1599   2.873.99   
61      1119216    299596                 1  PAC1044   2.780.99   
64      1119219    299596                 1  PAC1583   2.316.99   
67      1119223    294806                 1  APP1849   2.558.99   
69      1119226    294806                 1  APP1864   2.797.99   
...         ...       ...               ...      ...        ...   
293407  1649112    526844                 1  APP2491   1.329.00   
293416  1649124    526852                 1  APP2255   1.503.00   
293668  1649611    527121                 1  APP2489   1.099.00   
293728  1649733    527181                 1  APP2489   1.099.00   
293862  1649999    452946                 1  APP2075   2.999.00   

                      date new_unit_price  
59     2017-01-01 09:53:11       2.873.99  
61     2017-01-01 09:55:45       2.780.99  
64     2017-01-01 10:00:17       2.316.99  
67     2017-01-01 10:09:15       2.558.99  
69     2017-01-01 10:15:14       2.797.99  
...                    ...            ...  
293407 2018-03-14 10:42:06       1.329.00  
293416 2018-03-14 10:54:29       1.503.00  
293668 2018-03-14 11:59:58       1.099.00  
293728 2018-03-14 12:17:40       1.099.00  
293862 2018-03-14 13:03:33       2.999.00  

[19080 rows x 7 columns]

## I have to delete all the id_orders related to the remaining 19080 new_unit_price with two decimal dots

In [184]:
ord_l_df_cut = ord_l_df_merge.copy()

In [186]:
ord_l_df_cut.shape[0]

293983

In [187]:
# find the order_id of the relative corrupted prices:
two_dot_order_ids_list = ord_l_df_cut.loc[ord_l_df_cut.new_unit_price.astype(str).str.contains("\d+\.\d+\.\d+"), "id_order"]#.unique()
print(type(two_dot_order_ids_list))
print(len(two_dot_order_ids_list))

<class 'pandas.core.series.Series'>
19080


In [188]:
# delete the rows that still have two decimal dots in new_unit_price:
ord_l_df_cut = ord_l_df_cut.loc[~ord_l_df_cut.id_order.isin(two_dot_order_ids_list)]

In [209]:
ord_l_df_cut.shape

(233339, 7)

In [190]:
293983 - 233339

60644

In [192]:
ord_l_df_cut["new_unit_price"] = pd.to_numeric(ord_l_df_cut["new_unit_price"])

In [204]:
ord_l_df_cut.info()

<class 'pandas.core.frame.DataFrame'>
Index: 233339 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                233339 non-null  int64         
 1   id_order          233339 non-null  int64         
 2   product_quantity  233339 non-null  int64         
 3   sku               233339 non-null  object        
 4   unit_price        233339 non-null  object        
 5   date              233339 non-null  datetime64[ns]
 6   new_unit_price    233339 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(3), object(2)
memory usage: 14.2+ MB


# FINAL DATAFRAME: order_lines -> ord_l_df_cut:
## instead of deleting 77733 like Vasil, I deleted 60644 rows

In [210]:
ord_l_df_cut.shape

(233339, 7)

In [202]:
print('Eu: ', round((233339*100)/293983, 2), '  Vasil: ', round((216250*100)/293983,2) )

Eu:  79.37   Vasil:  73.56


In [194]:
ord_l_df_cut.head(10)

id  id_order  product_quantity      sku unit_price  \
0  1119109    299539                 1  OTT0133      18.99   
1  1119110    299540                 1  LGE0043     399.00   
2  1119111    299541                 1  PAR0071     474.05   
3  1119112    299542                 1  WDT0315      68.39   
4  1119113    299543                 1  JBL0104      23.74   
5  1119114    295310                10  WDT0249     231.79   
6  1119115    299544                 1  APP1582   1.137.99   
7  1119116    299545                 1  OWC0100      47.49   
8  1119119    299546                 1  IOT0014      18.99   
9  1119120    295347                 1  APP0700      72.19   

                 date  new_unit_price  
0 2017-01-01 00:07:19           18.99  
1 2017-01-01 00:19:45          399.00  
2 2017-01-01 00:20:57          474.05  
3 2017-01-01 00:51:40           68.39  
4 2017-01-01 01:06:38           23.74  
5 2017-01-01 01:14:27          231.79  
6 2017-01-01 01:17:21         1137.99  
7 2017-01-01 01:46:16           47.49  
8 2017-01-01 01:50:34           18.99  
9 2017-01-01 01:54:11           72.19

## 2 countercheck

In [151]:
ord_l_df_merge[ (ord_l_df_merge['unit_price'].str.contains("\d+\.\d+\.\d+") &
                            (ord_l_df_no_dupl['product_quantity']==1) )]

id  id_order  product_quantity      sku unit_price  \
6       1119115    299544                 1  APP1582   1.137.99   
11      1119126    299549                 1  PAC0929   2.565.99   
15      1119131    299553                 1  APP1854   3.278.99   
43      1119195    299582                 1  PAC0961   2.616.99   
105     1119294    299629                 1  QNA0175   2.054.99   
...         ...       ...               ...      ...        ...   
293838  1649954    527286                 1  APP2490   1.159.00   
293887  1650045    527321                 1  PAC2148   3.497.00   
293889  1650050    527324                 1  PAC2117   3.075.00   
293911  1650088    527342                 1  APP2492   1.329.00   
293936  1650135    527364                 1  APP2491   1.329.00   

                      date new_unit_price  
6      2017-01-01 01:17:21        1137.99  
11     2017-01-01 02:07:42        2565.99  
15     2017-01-01 02:14:47        3278.99  
43     2017-01-01 08:54:00        2616.99  
105    2017-01-01 11:39:13        2054.99  
...                    ...            ...  
293838 2018-03-14 12:55:14        1165.99  
293887 2018-03-14 13:10:15        3503.99  
293889 2018-03-14 13:10:56         3075.0  
293911 2018-03-14 13:24:51        1335.99  
293936 2018-03-14 13:36:13         1329.0  

[17096 rows x 7 columns]

In [146]:
ord_l_df_merge[ord_l_df_merge['id_order'].isin(corrupted_ids)]

id  id_order  product_quantity      sku unit_price  \
6       1119115    299544                 1  APP1582   1.137.99   
11      1119126    299549                 1  PAC0929   2.565.99   
15      1119131    299553                 1  APP1854   3.278.99   
43      1119195    299582                 1  PAC0961   2.616.99   
105     1119294    299629                 1  QNA0175   2.054.99   
...         ...       ...               ...      ...        ...   
293838  1649954    527286                 1  APP2490   1.159.00   
293887  1650045    527321                 1  PAC2148   3.497.00   
293889  1650050    527324                 1  PAC2117   3.075.00   
293911  1650088    527342                 1  APP2492   1.329.00   
293936  1650135    527364                 1  APP2491   1.329.00   

                      date new_unit_price  
6      2017-01-01 01:17:21        1137.99  
11     2017-01-01 02:07:42        2565.99  
15     2017-01-01 02:14:47        3278.99  
43     2017-01-01 08:54:00        2616.99  
105    2017-01-01 11:39:13        2054.99  
...                    ...            ...  
293838 2018-03-14 12:55:14        1165.99  
293887 2018-03-14 13:10:15        3503.99  
293889 2018-03-14 13:10:56         3075.0  
293911 2018-03-14 13:24:51        1335.99  
293936 2018-03-14 13:36:13         1329.0  

[17096 rows x 7 columns]

## 1 countercheck:

In [171]:
ord_l_df.loc[ord_l_df['id_order']==452946]

id  id_order  product_id  product_quantity      sku unit_price  \
293862  1649999    452946           0                 1  APP2075   2.999.00   
293863  1650000    452946           0                 1  BEL0371      26.99   
293864  1650001    452946           0                 1  SAT0054      69.99   
293865  1650003    452946           0                 1  APP1214      81.99   

                      date  
293862 2018-03-14 13:03:33  
293863 2018-03-14 13:03:33  
293864 2018-03-14 13:03:33  
293865 2018-03-14 13:04:17

In [213]:
# once I drop all the duplicates, including the first occurrence with keep=False, then
# I get the df I wanted:
#ord_l_df_no_dupl = ord_l_df.drop_duplicates(subset=['id_order'], keep=False)
ord_l_df_no_dupl.info()
ord_l_df_no_dupl.loc[ord_l_df_no_dupl['id_order']==452946]

<class 'pandas.core.frame.DataFrame'>
Index: 156141 entries, 0 to 293982
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                156141 non-null  int64         
 1   id_order          156141 non-null  int64         
 2   product_quantity  156141 non-null  int64         
 3   sku               156141 non-null  object        
 4   unit_price        156141 non-null  object        
 5   date              156141 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(3), object(2)
memory usage: 12.4+ MB


Empty DataFrame
Columns: [id, id_order, product_quantity, sku, unit_price, date]
Index: []

In [214]:
ord_l_df_cut.info()

<class 'pandas.core.frame.DataFrame'>
Index: 233339 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                233339 non-null  int64         
 1   id_order          233339 non-null  int64         
 2   product_quantity  233339 non-null  int64         
 3   sku               233339 non-null  object        
 4   unit_price        233339 non-null  object        
 5   date              233339 non-null  datetime64[ns]
 6   new_unit_price    233339 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(3), object(2)
memory usage: 14.2+ MB


In [222]:
two_dots = ord_l_df_cut.unit_price.astype(str).str.contains("\d+\.\d+\.\d+")

ord_l_df_cut.loc[two_dots][['unit_price', 'new_unit_price']].head(50)

#ord_l_df_cut.loc[two_dots].isna().sum()

unit_price  new_unit_price
6      1.137.99         1137.99
11     2.565.99         2565.99
15     3.278.99         3278.99
43     2.616.99         2616.99
105    2.054.99         2054.99
111    2.172.99         2172.99
131    2.050.99         2050.99
155    1.426.99         1426.99
159    1.351.99         1351.99
163    1.426.99         1426.99
164    1.348.99         1348.99
174    1.845.99         1845.99
187    1.019.00         1019.00
201    2.098.99         2098.99
225    3.318.99         3318.99
230    1.584.99         1584.99
240    2.133.99         2133.99
242    2.626.99         2626.99
264    2.698.99         2698.99
270    1.986.99         1986.99
279    1.398.99         1398.99
281    3.830.99         3830.99
303    6.029.99         6029.99
305    1.986.99         1986.99
328    1.688.99         1688.99
342    1.024.99         1024.99
356    3.198.99         3198.99
380    1.776.99         1776.99
386    1.986.99         1986.99
418    2.089.99         2089.99
459    1.114.99         1114.99
465    1.922.99         1922.99
471    1.649.99         1649.99
498    1.348.99         1348.99
501    1.024.99         1024.99
508    1.998.99         1998.99
524    1.794.99         1794.99
595    1.598.99         1598.99
615    1.193.99         1193.99
747    1.688.99         1688.99
755    1.949.99         1949.99
860    2.209.99         2209.99
941    1.584.99         1584.99
977    1.259.99         1259.99
1025   3.984.99         3984.99
1087   1.612.99         1612.99
1120   4.038.99         4038.99
1122   3.441.99         3441.99
1139   1.998.99         1998.99
1188   1.370.99         1370.99